feat: bb uses goblin

codygunton opened the pull request:
"vertical slice" of goblin integrating with ACIR and BB.

Beginning of goblin ultra honk interface for bb. Can only run a basic test (assert_statement) for now and is not intended for use outside of the CI check that it's still working, for now.
- adds bb and bb.js command `prove_and_verify_goblin`. Adds bb.js bindings for goblin prove/verify
- modifies ACIR `dsl` folder to be able to take a goblin builder
- adds CI calls to bb.js and bb that use `prove_and_verify_goblin`
- allowing ability to load grumpkin SRS through memory, needed for bb and (especially) bb.js. This allows an alternate source of points to be used other than the default file-based grumpkin loader, which mostly only works in dev (though revisit: could this work for native bb?)

codygunton committed:
Shared notes

codygunton committed:
Pre-convo work

codygunton committed:
Revert false start from yesterday

codygunton committed:
It builds

codygunton committed:
Notes etc from chat at EOD Friday

kevaundray commented on barretenberg/cpp/src/barretenberg/dsl/acir_proofs/major-todos.md:
```diff
@@ -0,0 +1,36 @@
+
+  Make acir_format and acir_proofs tests work
+
+  Stetch: try to understand what is_recursive flag should do
```
We allow two proving systems to be created, once of them is "snark friendly" meaning that it uses PedersenBlake in the transcript and the other is solidity friendly meaning that it uses keccak in the transcript.

When creating a proof that we will want to verify in a another circuit, we use the is_recursive flag. 

See https://hackmd.io/@6iQDuIePQjyYBqDChYw_jg/SyLHcKsSa for a few more details

codygunton replied:
Thanks for explaining. Our plans is to do violence to the code to get something working in the non-recursive case this week then check in with you and Maxim (prob next year?) to look things over together and rework to something non-disgusting :)

kevaundray replied:
Oh god :)

kevaundray replied:
Bear in mind that if a release is made, then this code will make its way into Noir user's hands -- haven't looked over the changes, though it seems that the solidity verifier contract may no longer work for them if this were to be merged?

codygunton replied:
Ok don't worry, we just want to get it working in this branch, won't merge

codygunton committed:
Put stuff in AcirComposer; everything builds.

codygunton committed:
Immunotherapy break

Co-authored-by: ledwards2225 <ledwards2225@users.noreply.github.com>

ledwards2225 committed:
wip

ledwards2225 committed:
building and running up to eccvm

ledwards2225 committed:
woops

ledwards2225 committed:
getting through proof construction

codygunton committed:
Add comment

codygunton committed:
No composer member

codygunton committed:
Should work but doesn't

codygunton committed:
Proofs constructed; ver fails

codygunton committed:
Try just Ultra

codygunton committed:
Test "passes"

codygunton committed:
Prune and update comments.

codygunton committed:
More touching up

codygunton committed:
Final touch-up before split

ludamad commented on barretenberg/acir_tests/flows/prove_and_verify.sh:
```diff
@@ -5,4 +5,5 @@ VFLAG=${VERBOSE:+-v}
-$BIN prove_and_verify $VFLAG -c $CRS_PATH -b ./target/acir.gz
\ No newline at end of file
+$BIN prove_and_verify $VFLAG -c $CRS_PATH -b ./target/acir.gz
+# lldb-16 -- $BIN prove_and_verify $VFLAG -c $CRS_PATH -b ./target/acir.gz
```
can be removed

codygunton committed:
Add proof_ assignment

codygunton committed:
Separate into clean function.

codygunton committed:
Clean up more

codygunton committed:
More cleanup

codygunton committed:
Add note

codygunton committed:
More cleanup and notes

ludamad0 committed:
feat: run prove_then_verify_goblin in CI

ledwards2225 committed:
templated necessary acir format fctns

ledwards2225 committed:
one more template fctn

ledwards2225 committed:
yet another

ledwards2225 committed:
holy fuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuuk

ledwards2225 committed:
simplify pnvGoblin plus debug

ledwards2225 committed:
new method properly accounts for noir bug

ledwards2225 committed:
stable state

ledwards2225 committed:
builder wire offset method works

ledwards2225 committed:
debug utility uses labels and db relation passes

ledwards2225 committed:
proof over goblin circuit verifies!

ledwards2225 committed:
quick cleanup pass

ledwards2225 committed:
more cleanup

ledwards2225 committed:
fix conversion issue plus goblin cleanup

ledwards2225 committed:
provide builder with goblin owned op queue

ledwards2225 committed:
separate goblin hacks for acir

ledwards2225 committed:
accum prove and verify returned to original state

ledwards2225 committed:
typename

ludamad0 committed:
fix: creating too few vars

ledwards2225 committed:
woops, actually fix rec verifier this time

ludamad0 committed:
fix: creating too few vars

ludamad committed:
feat: bb.js calling goblin ultrahonk (#3690)

- Upload grumpkin, start to use it
- Add more utilities for CRS grumpkin management to TS and C++
- Integrate 6_array tests into CI

---------

Co-authored-by: codygunton <codygunton@gmail.com>
Co-authored-by: ledwards2225 <l.edwards.d@gmail.com>
Co-authored-by: ledwards2225 <ledwards2225@users.noreply.github.com>
Co-authored-by: ludamad <adam@aztecprotocol.com>

ludamad committed:
Merge branch 'master' into cg-lde/expose-goblin

ludamad0 committed:
Get ci working

ludamad commented on barretenberg/cpp/src/barretenberg/bb/main.cpp:
```diff
@@ -31,23 +32,46 @@ acir_proofs::AcirComposer init(acir_format::acir_format& constraint_system)
 
 acir_format::WitnessVector get_witness(std::string const& witness_path)
 {
+    // WORKTODO(NEW_CONSTRAINTS): opqueue data is now being extracted here?
```
Should this get an issue?

ledwards2225 replied:
outdated, removed

ludamad commented on barretenberg/cpp/src/barretenberg/dsl/acir_format/acir_format.cpp:
```diff
@@ -1,19 +1,60 @@
+{
+    builder.variables[0] = 0; // WORKTODO(ZEROINDEX): This the constant 0 hacked in. Bad.
     for (size_t i = 0; i < witness.size(); ++i) {
+        // WORKTODO(ZEROINDEX): The i+1 accounts for the fact that 0 is added as a constant in variables in the UCB
```
Should this get an issue?

ledwards2225 replied:
done

ludamad commented on barretenberg/cpp/src/barretenberg/dsl/acir_format/acir_format.cpp:
```diff
@@ -1,19 +1,60 @@
+// be populated with zeros). The actual entries of the variables vector are populated in "read_witness"
+template <typename Builder> void add_public_vars(Builder& builder, acir_format const& constraint_system)
 {
+    // WORKTODO(ZEROINDEX): i = 1 acounting for const 0 in first position?
```
Should this get an issue?

ledwards2225 replied:
Done

ludamad commented on barretenberg/cpp/src/barretenberg/dsl/acir_format/acir_format.cpp:
```diff
@@ -104,22 +146,26 @@ void build_constraints(Builder& builder, acir_format const& constraint_system, b
-            builder.set_recursive_proof(proof_output_witness_indices);
+    // WORKTODO: disable these for UGH for now since we're not yet dealing with proper recursion
+    if constexpr (IsGoblinBuilder<Builder>) {
+        info("WARNING: this circuit contains recursion_constraints!");
```
Should this get an issue?

ledwards2225 replied:
done

ludamad commented on barretenberg/cpp/src/barretenberg/dsl/acir_format/acir_to_constraint_buf.hpp:
```diff
@@ -265,6 +265,7 @@ acir_format circuit_buf_to_acir_format(std::vector<uint8_t> const& buf)
     auto circuit = Circuit::Circuit::bincodeDeserialize(buf);
 
     acir_format af;
+    // WORKTODO(ZEROINDEX): this +1 seems to be accounting for the const 0 at the first index in variables
```
Should this get an issue?

ledwards2225 replied:
done

ludamad commented on barretenberg/cpp/src/barretenberg/dsl/acir_proofs/acir_composer.cpp:
```diff
@@ -75,6 +77,28 @@ std::vector<uint8_t> AcirComposer::create_proof(acir_format::acir_format& constr
+    // Correct for the addition of const variables in the builder constructor
+    acir_format::apply_wire_index_offset(goblin_builder_);
+
+    // WORKTODO: Add some arbitrary op gates to ensure the associated polynomials are non-zero
```
Should this get an issue?

ledwards2225 replied:
outdated

ludamad commented on barretenberg/cpp/src/barretenberg/dsl/acir_proofs/acir_composer.hpp:
```diff
@@ -1,18 +1,21 @@
+/**
+ * @brief A class responsible for marshalling construction of keys and prover and verifier instances used to prove
+ * satisfiability of circuits written in ACIR.
+ * @todo WORKTODO: This reflects the design of Plonk. Perhaps we should author new classes to better reflect the
```
Should this get an issue?

ledwards2225 replied:
done

ludamad commented on barretenberg/cpp/src/barretenberg/dsl/acir_proofs/major-todos.md:
```diff
@@ -0,0 +1,36 @@
+Major TODOS
+acir_format::Composer (and other types) become GUH
+? stretch update create_circuit to use Goblin gates (build_contraints and circuit_buf_to_acir_format in particular?)
+Update AcirComposer
```
Should this be deleted?

ledwards2225 replied:
deleted

ludamad commented on barretenberg/cpp/src/barretenberg/goblin/goblin.cpp:
```diff
@@ -0,0 +1,2 @@
+// NB: This file is here so that goblin_objects will be created
```
Good with just removing the WORKTODO here for now. I am itching after the new year to mass split templates into cpp files + instantiations

ledwards2225 replied:
done

ludamad commented on barretenberg/cpp/src/barretenberg/goblin/goblin.hpp:
```diff
@@ -13,30 +15,52 @@ namespace barretenberg {
 
+    // using GUHFlavor = proof_system::honk::flavor::Ultra;
+    // using GoblinUltraCircuitBuilder = proof_system::UltraCircuitBuilder;
+    using GUHFlavor = proof_system::honk::flavor::GoblinUltra;                 // GUHFLAG
```
Can clean up this GUHFLAG stuff now

ledwards2225 replied:
cleaned

ludamad commented on barretenberg/cpp/src/barretenberg/proof_system/circuit_builder/goblin_ultra_circuit_builder.hpp:
```diff
@@ -20,6 +20,8 @@ template <typename FF> class GoblinUltraCircuitBuilder_ : public UltraCircuitBui
     static constexpr size_t DEFAULT_NON_NATIVE_FIELD_LIMB_BITS =
         UltraCircuitBuilder_<arithmetization::UltraHonk<FF>>::DEFAULT_NON_NATIVE_FIELD_LIMB_BITS;
 
+    size_t num_vars_added_in_constructor; // needed in constructing circuit from acir
```
nit: could have an = 0 for sanity

ledwards2225 replied:
done

ludamad commented on barretenberg/cpp/src/barretenberg/proof_system/circuit_builder/ultra_circuit_builder.hpp:
```diff
@@ -56,6 +56,8 @@ class UltraCircuitBuilder_ : public CircuitBuilderBase<typename Arithmetization:
     // number of gates created per non-native field operation in process_non_native_field_multiplications
     static constexpr size_t GATES_PER_NON_NATIVE_FIELD_MULTIPLICATION_ARITHMETIC = 7;
 
+    size_t num_vars_added_in_constructor; // needed in constructing circuit from acir
```
nit: could have an = 0 for sanity just in case code gets deleted

ledwards2225 replied:
done

ludamad commented on barretenberg/cpp/src/barretenberg/proof_system/debug_utility.hpp:
```diff
@@ -0,0 +1,43 @@
+#pragma once
+
+#include "barretenberg/common/log.hpp"
```
nit: usually there's a better name than `X_utility`. But to be clear, it's holidays, get'er in

ledwards2225 replied:
changed to `instance_inspector` - useful for now but may delete in the near future

ludamad commented on barretenberg/cpp/src/barretenberg/ultra_honk/ultra_composer.hpp:
```diff
@@ -1,5 +1,8 @@
 #pragma once
 #include "barretenberg/flavor/flavor.hpp"
+#include "barretenberg/plonk/proof_system/proving_key/proving_key.hpp"           // WORKTODO
```
Should this get an issue?

ledwards2225 replied:
these were unneeded - deleted

ludamad commented on barretenberg/cpp/src/barretenberg/ultra_honk/ultra_verifier.cpp:
```diff
@@ -138,6 +138,7 @@ template <typename Flavor> bool UltraVerifier_<Flavor>::verify_proof(const plonk
 
     // If Sumcheck did not verify, return false
     if (sumcheck_verified.has_value() && !sumcheck_verified.value()) {
+        info("UltraVerifier: Sumcheck failed.");
```
stray info

ledwards2225 replied:
I'd like to keep this lil guy around for now

ludamad committed:
Update index.ts

ludamad commented on barretenberg/ts/src/crs/node/ignition_files_crs.ts:
```diff
@@ -14,9 +14,10 @@ function getCurrentDir() {
+ * The path to our SRS object, assuming that we are in e.g. barretenberg/ts/dest/node/crs/node folder.
  */
-export const SRS_DEV_PATH = getCurrentDir() + '/../../../cpp/srs_db/ignition/monomial';
+export const SRS_DEV_PATH = getCurrentDir() + '/../../../../../cpp/srs_db/ignition/monomial';
```
the old path actually never found the file, always downloaded

ledwards2225 committed:
update WORKTODOs

ledwards2225 committed:
Merge branch 'cg-lde/expose-goblin' of github.com:AztecProtocol/aztec-packages into cg-lde/expose-goblin

ledwards2225 committed:
respond to feedback and clean up

kevaundray commented:
Wait why has this now been merged?

see: https://github.com/AztecProtocol/aztec-packages/pull/3636#discussion_r1421612722

ledwards2225 commented:
> Wait why has this now been merged?
> 
> see: [#3636 (comment)](https://github.com/AztecProtocol/aztec-packages/pull/3636#discussion_r1421612722)

Not to worry! This still isn't hooked up to anything real, it just runs a few new tests using Goblin. Sorry - poor communication on our part. When Cody said "this wont be merged" we were still thinking it wouldn't be. Since then we decided to do something more isolated very much for the purpose of getting it merged
